PDCL-6011 Migrate configuration view to react spectrum 3. Supports paging for edge configuration selection.

Aaronius opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description

<!--- Describe your changes in detail -->

## Related Issue

<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->

## Motivation and Context

<!--- Why is this change required? What problem does it solve? -->

## Screenshots (if appropriate):

## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] All tests pass and I've made any necessary test changes.
- [ ] I've updated the schema in extension.json or no changes are necessary.

Aaronius commented on .eslintrc.js:
```diff
@@ -10,8 +10,10 @@ module.exports = {
+  plugins: ["unused-imports", "ban", "prettier", "testcafe"],
   rules: {
+    "unused-imports/no-unused-imports": "error",
+    "unused-imports/no-unused-vars": "error",
```
This automatically removed unused imports. I got tired of doing it manually.

Aaronius commented on src/view/actions/setConsent.jsx:
```diff
@@ -310,8 +309,7 @@ const ConsentObject = ({ value, index }) => {
             dataTestIdPrefix="gdprContainsPersonalData"
             name={`consent[${index}].gdprContainsPersonalData`}
-            label="GDPR Contains Personal Data"
-            description="Does the event data associated with this user contain personal data?"
```
I changed FieldDescriptionAndError to derive its width based off its only child, rather than having to pass `width` to both `FieldDescriptionAndError` and the child of `FieldDescriptionAndError`.  I did this before you created RadioGroupWithDataElement. When I rebased on top of your changes, RadioGroupWithDataElement was feeding multiple children to FieldDescriptionAndError, so that was causing an error.

I debated on how to solve it, but I thought the UX was kind of cluttered with a description at the RadioGroupWithDataElement level:

![image](https://user-images.githubusercontent.com/210820/125351079-dcec5f80-e31c-11eb-8263-1b0e80ac2fed.png)

So I removed FieldDescriptionAndError from RadioGroupWithDataElement and just made the labels be what the descriptions were.

Aaronius replied:
👍  Thanks. I'll update the docs.

Aaronius replied:
@jonsnyder When you mentioned that this change requires a change to the documentation, I believe you were thinking that the labels were referenced in the documentation and those references needed be changed. Just so you're aware, I didn't see any label references from the documentation (I assume they would be [here](https://experienceleague.adobe.com/docs/experience-platform/edge/extension/action-types.html?lang=en#set-consent), so I don't think anything needs to be changed. If you're aware of where they may be referenced or if you were thinking about something different when you made your comment, let me know.

Aaronius commented on src/view/errors/userReportableError.js:
```diff
@@ -18,19 +18,14 @@ governing permissions and limitations under the License.
-      // we can still see its stack trace.
-      console.error(originatingError);
-    }
-
```
See monitorForOriginatingErrors for why I removed this. TL;DR: if this UserReportableError is handled, it's confusing to have the originating error logged to the console.

Aaronius commented on src/view/utils/fetchFromEdge.js:
```diff
@@ -0,0 +1,88 @@
+    imsAccess,
+    productionHost: EDGE_HOST_PROD,
+    stagingHost: EDGE_HOST_STAGING
+  });
```
Previously we were just using the prod endpoint for edge configurations, but I figured we should be consistent with what we're doing for the aep schema endpoints and support both staging and production.

Aaronius commented on src/view/dataElements/xdmObject/components/sandboxSelector.jsx:
```diff
@@ -54,13 +53,11 @@ const SandboxSelector = ({
-            })
-          );
-          return undefined;
+          reportAsyncError(e);
```
I moved the creating of UserReportableError up to fetchSandboxes (and did so similarly in other places). It seemed to make the most sense there and reduced duplicate code.

Aaronius commented on src/view/configuration/utils/getPartsFromEnvironmentCompositeId.js:
```diff
@@ -10,22 +10,18 @@ OF ANY KIND, either express or implied. See the License for the specific languag
+  };
 };
 
-export default abortPreviousRequestsAndCreateNewSignal;
```
This diff is messed up; this wasn't a renamed file. `abortPreviousRequestsAndCreateNewSignal` was something I was using canceling requests in the XDM object data element. Notice that I had a "global" `abortController`, which was fine for the XDM object case, but it meant I couldn't re-use it for the  edge config selection in the configuration view. I decided to change it to a re-usable React hook, so the abortController would be scoped to the component using it. See `useAbortPreviousRequestsAndCreateSignal.jsx`.

Aaronius commented on test/functional/actions/sendEvent.spec.js:
```diff
@@ -156,6 +153,7 @@ test("returns full valid settings with decision scopes as data element", async (
   });
   await instanceNameField.selectOption("alloy2");
   await typeField.enterSearch("mytype1");
+  await typeField.pressEnterKey();
```
I upgraded React-Spectrum's ComboBox and their new version requires hitting enter to accept the text instead of just blurring.

Aaronius commented on test/functional/actions/resetEventMergeId.spec.js:
```diff
@@ -10,14 +10,10 @@ OF ANY KIND, either express or implied. See the License for the specific languag
 import createFixture from "../helpers/createFixture";
 
-const extensionViewController = createExtensionViewController(
-  "actions/resetEventMergeId.html"
```
The path here wasn't actually being used since I re-worked how tests are run. This means `createExtensionViewController` could just change to `extensionViewController`.

Aaronius commented on test/functional/configuration/configuration.spec.js:
```diff
@@ -10,180 +10,76 @@ OF ANY KIND, either express or implied. See the License for the specific languag
-  tokens: {
-    imsAccess: "Badtoken"
-  }
-};
```
I moved this default init info into `extensionViewController`, which simplified some things.

Aaronius commented on test/functional/configuration/helpers/viewSelectors.js:
```diff
@@ -0,0 +1,129 @@
+  });
+}
+
+export { instances };
```
I split these selectors into a separate file because (1) it was getting rather large and (2) I plan to re-use them in a spec file that tests edge configuration selection.

jonsnyder commented on src/view/configuration/edgeConfigurationsSection.jsx:
```diff
@@ -0,0 +1,397 @@
+    }
+
+    if (!isSuccessfullyPopulatedForSelectInputMethod) {
+      instanceValues.edgeConfigSelectInputMethod = await getSelectInputMethodStateForNewInstance(
```
I don't understand this part. If this is the second instance there is no need to fetch configs. Indeed, it will return configs for the wrong IMS org.

jonsnyder commented on src/view/configuration/edgeConfigurationsSection.jsx:
```diff
@@ -0,0 +1,397 @@
+        candidateInstance =>
+          getEdgeConfigIdFromInstance(candidateInstance) === edgeConfigId
+      );
+      const isDuplicate = firstInstanceWithSameEdgeConfigId !== instance;
```
Clever!

jonsnyder commented on src/view/configuration/edgeConfigurationsSection.jsx:
```diff
@@ -0,0 +1,397 @@
+          >
+            Enter values
+          </Radio>
+        </RadioGroup>
```
Only the first instance should have this radio group. We cannot fetch the data streams for the second instance because the IMS org will be different.

jonsnyder reviewed (changes requested):
Lots of good stuff in here! My biggest concern is showing the data streams input method selector on the second instance.

Aaronius commented:
Thanks for the feedback! I've addressed your comments. Take a look at the last commit in particular. I created a method called `getMergedBridgeMethod` that's used for the various bridge methods. One side effect of this is that it makes `getSettings` asynchronous. Launch supports an asynchronous `getSettings`, and we could end up needing it eventually, so I decided to roll with it. It does change spectrum3ExtensionView.jsx, which might conflict with what you're working on.

jonsnyder commented:
I found some issues while testing the UI:

1. The descriptions on the privacy options are missing
2. I think the onBeforeEventSend label should be "On before event send callback". For consistency we have not used the property name as the label before.
3. Required field asterisks are not consistent. (IMS Org Id, Name, Edge Domain, Default Consent, Edge base path, Consistency between datastreams select and form, download link qualifier)
4. "Restore to default" and "Test Regex" buttons are not clearly buttons. In the old UI, the hover state for these showed the outline of a button.
5. Too much space between Enable Click Collection checkbox and Download Link Qualifier
6. In extension.json, downloadLinkQualifier has a minimum length of 1, but there is no such validation in the UI

jonsnyder commented on src/view/configuration/configuration.jsx:
```diff
@@ -79,33 +97,11 @@ const getInitialValues = async ({ initInfo }) => {
   if (instancesSettings) {
     instancesInitialValues = await Promise.all(
       instancesSettings.map(async (instanceSettings, instanceSettingsIndex) => {
```
No longer need this async.

jonsnyder reviewed (approved):
Looks good. Thanks!

Aaronius commented on src/view/components/formikReactSpectrum3/radioGroupWithDataElement.jsx:
```diff
@@ -38,7 +38,6 @@ const RadioGroupWithDataElement = ({
   name,
   children,
   dataElementDescription,
-  width,
```
Having `width` on this component didn't really do what was expected because we were forwarding on the width to the ReactSpectrumRadioGroup and the data element textfield, which is indented. Meaning the total width of the component ends up being larger than defined width. Everywhere we're currently using RadioGroupWithDataElement seems to look better without a constrained width as well. Because of these things, I just decided to remove the prop entirely and if we want to add support for width later, we'll need to be smarter about how we accomplish it.

Aaronius commented on src/view/components/formikReactSpectrum3/radioGroupWithDataElement.jsx:
```diff
@@ -141,15 +140,16 @@ const RadioGroupWithDataElement = ({
                 }
-                width={width}
+                width="size-5000"
+                isRequired
```
When the "data element" radio is selected, this textfield becomes required, so I think it makes sense to mark the textfield as required. The required indicator only shows up if we have a label though, which is why I switched the prop from `aria-label` to `label`. I added labels to other textfields that didn't have labels to be consistent.

Aaronius commented on src/view/configuration/dataCollectionSection.jsx:
```diff
@@ -146,60 +146,62 @@ const DataCollectionSection = ({ instanceFieldName }) => {
+              </Flex>
+            </FieldSubset>
+          )}
+        </div>
```
Wrapped in a div because FormElementContainer was adding space between the checkbox and regex field, causing the additional spacing you mentioned.

Aaronius commented on src/view/dataElements/xdmObject/components/objectEdit.jsx:
```diff
@@ -55,7 +55,8 @@ const ObjectEdit = ({ fieldName }) => {
             <TextField
               data-test-id="valueField"
               name={`${fieldName}.value`}
-              label="Data element providing array"
```
Oops. This should have said "object".

Aaronius commented:
Thanks for the finds, @jonsnyder! I've made the updates. Regarding marking fields as required, I talked to the Spectrum folk and they said:

> No need to mark it as required if the user can't not fill it out.

I decided to go with that approach.

Aaronius commented:
Thanks for the finds, @jonsnyder! I've made the updates. Regarding marking fields as required, I talked to the Spectrum folk and they said:

> No need to mark it as required if the user can't not fill it out.

I decided to go with that approach.

Aaronius commented:
@jonsnyder I'm going to merge this, but if you review the latest changes and have feedback, I'm happy to address them in a later PR. I have an upcoming PR for removing React-Spectrum 2 stuff, so making additional changes in this branch is making it more difficult to keep that other branch properly rebased.
